### ISSUE TITLE
Workspaces backward object relations

### ISSUE THREAD

**Issue by JPaja** [Washi1337/AsmResolver#164]
## Problem Description
Object relations are one way so there is no possibility of backward traversal

## Proposal
Change structure of ObjectRelation such that it has both way types 

```csharp
ObjectRelation<MethodDefinition, MethodReference> MethodReferenceRelation
```
where 
```csharp
node1.Subject is MethodDefinition
node2.Subject is MethodReference
```
and API would look like
```csharp
node1.AddRelation(MethodReferenceRelation, node2);

node2.GetAllBackwardRelatedNodes().Contains(node1) //true
node2.GetBackwardRelatedNodes(MethodReferenceRelation).Contains(node1) //true
```

## Alternatives
Do not change ObjectRelation class and just fill some collection of nodes inside WorkspaceIndexNode on each AddRelation
that would represent backward nodes.

**Comment by Washi1337** [comment]
Note: If we are going to keep track of incoming edges for every node, it is probably going to be worth it to fully extract the neighbors collection into a list of edges (with an actual edge type), rather than just a collection of adjacent nodes. This is useful for both syncing and lookup of the backwards related nodes. Similar code was written for this in the [adjacency collection implementation of Echo](https://github.com/Washi1337/Echo/blob/master/src/Core/Echo.ControlFlow/Collections/AdjacencyCollection.cs).

**Comment by Washi1337** [comment]
Merged in 20ddbbacac9e06e8724d81446a29d6cd36e7069a
